transport/sqs/src/main/kotlin/SqsMessageReceiverFactory.kt

```diff
@@ -0,0 +1,125 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.eclipse.apoapsis.ortserver.transport.sqs
+
+import aws.sdk.kotlin.services.sqs.model.DeleteMessageRequest
+import aws.sdk.kotlin.services.sqs.model.GetQueueUrlRequest
+import aws.sdk.kotlin.services.sqs.model.MessageAttributeValue
+import aws.sdk.kotlin.services.sqs.model.ReceiveMessageRequest
+
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.runBlocking
+
+import org.eclipse.apoapsis.ortserver.config.ConfigManager
+import org.eclipse.apoapsis.ortserver.transport.Endpoint
+import org.eclipse.apoapsis.ortserver.transport.EndpointHandler
+import org.eclipse.apoapsis.ortserver.transport.Message
+import org.eclipse.apoapsis.ortserver.transport.MessageHeader
+import org.eclipse.apoapsis.ortserver.transport.MessageReceiverFactory
+import org.eclipse.apoapsis.ortserver.transport.RUN_ID_PROPERTY
+import org.eclipse.apoapsis.ortserver.transport.TOKEN_PROPERTY
+import org.eclipse.apoapsis.ortserver.transport.TRACE_PROPERTY
+import org.eclipse.apoapsis.ortserver.transport.json.JsonSerializer
+
+import org.slf4j.LoggerFactory
+
+private val logger = LoggerFactory.getLogger(SqsMessageReceiverFactory::class.java)
+
+internal val messageAttributeNames = listOf(TOKEN_PROPERTY, TRACE_PROPERTY, RUN_ID_PROPERTY)
+
+/**
+ * A [MessageReceiverFactory] implementation for AWS SQS.
+ * Also see https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/kotlin_sqs_code_examples.html.
+ */
+class SqsMessageReceiverFactory : MessageReceiverFactory {
+    override val name: String = SqsConfig.TRANSPORT_NAME
+
+    override fun <T : Any> createReceiver(
+        from: Endpoint<T>,
+        configManager: ConfigManager,
+        handler: EndpointHandler<T>
+    ) = runBlocking {
+        logger.info("Creating SQS receiver for endpoint '${from.configPrefix}'.")
+
+        val serializer = JsonSerializer.forClass(from.messageClass)
+        val config = SqsConfig.create(configManager)
+        val client = createSqsClient(config)
+
+        val queueResponse = client.getQueueUrl(GetQueueUrlRequest { queueName = config.queueName })
+
+        val queueUrl = checkNotNull(queueResponse.queueUrl) {
+            "The SQS queue URL for queue '$name' must not be null."
+        }
+
+        val receiveRequest = ReceiveMessageRequest {
+            this.queueUrl = queueUrl
+
+            // Opt-in to the message attributes to receive.
+            messageAttributeNames = org.eclipse.apoapsis.ortserver.transport.sqs.messageAttributeNames
+
+            // Enable "long polling" to eliminate empty responses, see
+            // https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/working-with-messages.html#setting-up-long-polling
+            waitTimeSeconds = 1
```

Comment:
The linked docs say that in most cases this can be set to 20 seconds, why did you chose 1? IIUC, if a message arrives within the 20 seconds the function would still return the message immediately. Or would it always wait the full time before returning anything?

Also, I was thinking if we should explicitly set [`maxNumberOfMessages`](https://sdk.amazonaws.com/kotlin/api/latest/sqs/aws.sdk.kotlin.services.sqs.model/-receive-message-request/index.html) to 1, even if it is the default, to explicitly state it here.

Reply:
> why did you chose 1?

Simply because I was unsure how much delay would be acceptable for ORT server.

> Or would it always wait the full time before returning anything?

My understanding so far was that if the poll interval is set to 20s, that would mean what messages might not be received for 20s. But reading thought [this](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-short-and-long-polling.html) again it says

> Long polling - [...] sending a response once at least one message is available

So it seems that using the recommended 20s should be ok, I'll make that change.

> Also, I was thinking if we should explicitly set [maxNumberOfMessages](https://sdk.amazonaws.com/kotlin/api/latest/sqs/aws.sdk.kotlin.services.sqs.model/-receive-message-request/index.html) to 1, even if it is the default, to explicitly state it here.

Should we rather do the opposite and increase this, to allow for more message to be delivered at once, possibly increasing performance?

Reply:
> Should we rather do the opposite and increase this, to allow for more message to be delivered at once, possibly increasing performance?

I think it's better to process messages immediately. For example, if a worker has finished, the orchestrator should start he next one immediately and not wait for other messages. And workers should not handle more than one message at a time to not limit parallelism (even though in Kubernetes the Kubernetes transport would typcially be used for "sending messages" from the orchestrator to the workers).

Reply:
> I think it's better to process messages immediately.

I agree, but I'm not reading the docs for [maxNumberOfMessages](https://sdk.amazonaws.com/kotlin/api/latest/sqs/aws.sdk.kotlin.services.sqs.model/-receive-message-request/index.html) so that it would always wait for that number of messages:

> "Amazon SQS never returns more messages than this value (however, fewer messages might be returned)."

> And workers should not handle more than one message at a time

IMO they can't away, as the signature of the `handler` function only allows one message to be passed at a time. So the question rather is: How quickly is it allowed to call the `handler` function in succession? In case of multiple received messages, I'm just looping over them, and calling the handler for each separately.

Reply:
This is transparent for the transport implementation. The handler function is blocking, so this means that it can handle the next message when it returns. For transport implementations based on message queues, this probably means that in order to process messages in parallel, a corresponding number of receivers will have to be deployed.

Reply:
> This is transparent for the transport implementation.

Exactly. The (my) question is not about processing messages in parallel or not, but about whether there's a difference (for the handler) to be called 10 times for 10 individually received messages, or 10 times for 10 messages received at once.

For the handler, this should not make a difference, as stated. But again my question remains, could it still be more efficient fro the AWS SDK perspective to allow receiving messages in bulk, even if they are passed one-by-one to the (blocking) handler?

Reply:
When the handler calls a worker which can take minutes for its execution, then I doubt that it has any measurable effect if the fetching of messages is done slightly more or less efficiently. It may make a difference if message processing was really fast. At least that would be my understanding.